deploy.sh

```diff
@@ -0,0 +1,54 @@
+#!/bin/sh
+
+set -e
+
+if ! git diff --cached --quiet; then
+  echo "error: refusing to deploy with changes staged in the index"
+  exit 1
+fi
+
+if ! git diff --quiet; then
+  echo "error: refusing to deploy with changes in the worktree"
+  exit 1
+fi
+
+
+TOP="$(git rev-parse --show-toplevel)"
+echo "Working from top level: $TOP"
+cd "$TOP"
+
+echo "Clearing index"
+git rm -r -q --cached .
+
+echo "Adding public/ files to index"
+find public -type f | while read -r SRC; do
+  HASH=$(cat $SRC | git hash-object -w --stdin --path .)
+  DEST=$(/bin/echo -n $SRC | sed -e 's;^public/\(.*\);\1;')
+
+  if [ -x "$SRC" ]; then
+    MODE=100755
+  else
+    MODE=100644
+  fi
+
+  /bin/echo -n "$MODE blob $HASH	$DEST" | git update-index --index-info
+
+  # Print progress indicator.
+  /bin/echo -n .
+done
+
+echo
+TREE=$(git write-tree)
+echo "Created tree object $TREE"
+
+echo "Creating commit object on gh-pages branch"
+ABBREV=$(git rev-parse --short HEAD)
+COMMIT=$(git commit-tree -m "Autogenerated public/ content from $ABBREV" -p gh-pages "$TREE")
+git branch -f gh-pages "$COMMIT"
+git log -1 gh-pages
+
+echo "Restoring index to prior state"
+git reset HEAD
+
+echo
+echo "Now run 'git push upstream gh-pages'"
```

Comment:
Obviously not everybody uses "upstream" as a branch name, but hopefully the required action is clear anyway. If people think it's necessary we can try to infer the right remote name from the output of `git remote -v`.